h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstDistance.java

```diff
@@ -0,0 +1,132 @@
+package water.rapids.ast.prims.advmath;
+
+import water.H2O;
+import water.MRTask;
+import water.fvec.*;
+import water.rapids.Env;
+import water.rapids.Val;
+import water.rapids.ast.AstPrimitive;
+import water.rapids.ast.AstRoot;
+import water.rapids.vals.ValFrame;
+import water.util.Log;
+
+/**
+ * Calculate Distance Metric between pairs of rows
+ */
+public class AstDistance extends AstPrimitive {
+  @Override
+  public String[] args() {
+    return new String[]{"ary", "x", "y", "measure"};
+  }
+
+  @Override
+  public int nargs() {
+    return 1 + 3; /* (distance X Y measure) */
+  }
+
+  @Override
+  public String str() {
+    return "distance";
+  }
+
+  @Override
+  public Val apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
+    Frame frx = stk.track(asts[1].exec(env)).getFrame();
+    Frame fry = stk.track(asts[2].exec(env)).getFrame();
+    String measure = stk.track(asts[3].exec(env)).getStr();
+    return computeCosineDistances(frx, fry, measure);
+  }
+
+  public Val computeCosineDistances(Frame references, Frame queries, String distanceMetric) {
```

Comment:
Does `references` and `queries` have a particular meaning? Or just names for two frames in which you will pair up subsequent rows and calculate the distance?